### ISSUE TITLE
Operators cannot be used with single-argument function shorthand in some cases

### ISSUE THREAD

**Issue by imm** [DanielXMoore/Civet#854]
It seems to me that I saw a discussion of a similar problem somewhere here a few months ago, but I couldn’t find it now and could have confused it with something, so I’ll write just in case, since I encountered this in practice.

At first I thought this only applied to the humanized NOT operator:
```js
collection.filter ({condition}) => condition // A lot of typing
collection.filter ({condition}) => not condition // Could it be shorter?
collection.filter &.condition  // good
collection.filter !&.condition // good
collection.filter not &.condition // error

↓↓↓

collection.filter(({condition}) => condition)
collection.filter(({condition}) => !condition)
collection.filter($ => $.condition)
collection.filter($1 => !$1.condition)
collection.filter(!$2 => $2.condition) // weird
```
But it turned out that for binary operators this approach does not work at all (presumably):
```js
arr.filter &.a and b    // good
arr.filter a and &.b    // bad
arr.filter a && &.b     // bad
arr.filter &.a and &.b  // bad

↓↓↓

arr.filter($ => $.a && b)
arr.filter(a && $1 => $1.b)
arr.filter(a && $2 => $2.b)
arr.filter($3 => $3.a && $4 => $4.b)
```

```js
arr.map &.a + b   // good
arr.map a + &.b   // bad
arr.map &.a + &.b // bad

operator sum
arr.map &.a sum b   // good
arr.map a sum &.b   // bad
arr.map &.a sum &.b // bad

↓↓↓

arr.map($ => $.a + b)
arr.map(a + $1 => $1.b)
arr.map($2 => $2.a + $3 => $3.b)


arr.map($4 => sum($4.a, b))
arr.map(sum(a, $5 => $5.b))
arr.map($6 => sum($6.a, $7 => $7.b))

```

It seems to me that this should either work or not work in all cases at once.

**Comment by edemaine** [comment]
`not &` omission is a bug.

The others are as designed (so far) : the function arrow goes where the `&` is. Without a rule like this, you don't know where to put the arrow; your other examples are inherently ambiguous. Unless you can come up with a better rule?

**Comment by edemaine** [comment]
Perhaps we should continue discussion in the #693 thread, which was about this issue.

Meanwhile, #883 fixes `not &` and adds documentation about what the current meaning of `&` is.

I'm familiar with Mathematica/Wolfram notation. Personally I've found it extremely counterintuitive, especially to read but also to write (despite knowing shell scripts where `&` presumably comes from). It also feels pretty ambiguous in terms of where the function boundary is, so you usually need at least one set of parentheses to delineate.

`&` has the advantage of usually not needing parentheses; it's effectively an open bracket that spans maximally to the right. I agree that it's limited (can't re-use the argument, can't put things to the left that aren't unary opreators), but we've struggled to find a more general notation that's any more concise than `($) => ...` which is already quite concise. If you come up with something that beats this and is more general than `&`, we'd love to hear it!

See also #85 for thoughts on re-using `&`.

I think there could be nice ways to handle the binary case, with the same limitations as `&`. Something like `x.reduce &1+&2`.
